modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Services/ServicesClient.cs

```diff
@@ -202,7 +217,127 @@ public IServicesClient WithServerKeepBinary()
+        {
+            /** Service name. */
+            private readonly string _svcName;
+            
+            /** Ignite services. */
+            private readonly ServicesClient _svcClient;
+
+            /** Flag of topology update progress. */
+            private volatile int _updateInProgress;
+
+            /** Time of the last update. */
+            private long _lastUpdateRequestTime;
+
+            /** Cluster topology version of the last update. */
+            private long _lastAffTop;
+            
+            /** Ids of the nodes with at least one service instance. */
+            private volatile IList<Guid> _nodes = new List<Guid>();
+
+            /// <summary>
+            /// Creates service topology holder.
+            /// </summary>
+            internal ServiceTopology(string name, ServicesClient svcClient)
+            {
+                _svcName = name;
+                _svcClient = svcClient;
+            }
+
+            /// <summary>
+            /// Asynchronously updates the topology.
+            /// </summary>
+            private async Task UpdateTopologyAsync()
+            {
+                if (Interlocked.CompareExchange(ref _updateInProgress, 1, 0) == 1)
+                    return;
+
+                var socket = _svcClient._ignite.Socket;
+
+                var topVer = socket.GetTopologyVersion();
+
+                var log = _svcClient._ignite.GetConfiguration().Logger;
+
+                var groupNodes = _svcClient._clusterGroup?.GetNodes();
+                
+                var top = await socket.DoOutInOpAsync(ClientOp.ServiceGetTopology,
+                    ctx => ctx.Writer.WriteString(_svcName),
+                    ctx =>
+                    {
+                        var cnt = ctx.Reader.ReadInt();
+
+                        var res = new List<Guid>(cnt);
+                        
+                        for (var i = 0; i < cnt; ++i)
+                            res.Add(BinaryUtils.ReadGuid(ctx.Reader.Stream));
+
+                        return res;
+                    }, (status, err) =>
+                    {
+                        log.Error("Failed to update topology of the service '" + _svcName + "'.", err);
+
+                        return _nodes;
+                    }).ConfigureAwait(false);
+                
+                _nodes = FilterTopology(top, groupNodes?.Select(n => n.Id).ToList());
+                
+                Interlocked.Exchange(ref _lastUpdateRequestTime, DateTime.Now.Ticks);
+                Interlocked.Exchange(ref _lastAffTop, topVer);
+
+                log.Debug("Topology of service '" + _svcName + "' has been updated. The " +
+                          "service instance nodes: " + string.Join(", ", top.Select(gid=>gid.ToString())) +
+                          ". Effective topology with the cluster group is: " + 
+                          string.Join(", ", _nodes.Select(gid=>gid.ToString())) + '.');
+
+                _updateInProgress = 0;
```

Comment:
Mixing `volatile` and `Interlocked` makes the code harder to reason about. It is considered a good practice to stick to one synchronization mechanism for a given field. Let's remove `volatile` and use `Interlocked.CompareExchange(ref _updateInProgress, 0, 0)` to read the value.

See https://stackoverflow.com/questions/24808291/reading-an-int-thats-updated-by-interlocked-on-other-threads